File: 19/19.py
import functools
from pathlib import Path
from pprint import pp


@functools.cache
def possible_variants(patterns: tuple[str], design: str) -> int:
    if design == "":
        return 1

    candidates = [pattern for pattern in patterns if design.startswith(pattern)]
    if not candidates:
        return 0

    return sum(
        possible_variants(patterns, design[len(pattern) :]) for pattern in candidates
    )


def parse(path: Path) -> tuple[tuple[str, ...], list[str]]:
    patterns, designs = path.read_text().split("\n\n")
    patterns = tuple(patterns.split(", "))
    designs = [design for design in designs.split("\n") if design]
    return patterns, designs


if __name__ == "__main__":
    path = Path(__file__).parent / "input.txt"
    patterns, designs = parse(path)
    print(sum(possible_variants(patterns, design) for design in designs))

File: 19/19.ex
defmodule Day19 do
  def solve do
    path = "input.txt"

    {:ok, {patterns, designs}} =
      path
      |> File.read!()
      |> String.split("\n\n", trim: true)
      |> parse()

    dbg()
    Enum.map(designs, &possible?(patterns, &1)) |> Enum.sum()
  end

  def parse([patterns, designs]) do
    patterns = String.split(patterns, ", ", trim: true)
    designs = String.split(designs)
    {:ok, {patterns, designs}}
  end

  def possible?(_patterns, ""), do: 1

  def possible?(patterns, design) do
    # Greedy search doesn't work, need to search all candidates
    case Enum.filter(patterns, fn pattern -> String.starts_with?(design, pattern) end) do
      nil ->
        0

      candidates ->
        Enum.map(
          candidates,
          fn candidate ->
            possible?(
              patterns,
              String.slice(design, String.length(candidate)..String.length(design))
            )
          end
        )
        |> Enum.sum()
    end
  end
end

IO.puts(Day19.solve())
